### PR TITLE
Add support for Sponsored Reserves (CAP33).

### PR DESCRIPTION
**Pull request by abuiles**
Extend the operation class with helpers that allow sponsoring reserves and also revoke sponsorships.

To start sponsoring reserves for an account use:
- `Operation.beginSponsoringFutureReserves`
- `Operation.endSponsoringFutureReserves`

To revoke a sponsorship after it has been created use any of the following helpers:

- `Operation.revokeAccountSponsorship`
- `Operation.revokeTrustlineSponsorship`
- `Operation.revokeOfferSponsorship`
- `Operation.revokeDataSponsorship`
- `Operation.revokeClaimableBalanceSponsorship`
- `Operation.revokeSignerSponsorship`

The following example contains a transaction which sponsors operations for an account and then revoke some sponsorships.

```js
const transaction = new StellarSdk.TransactionBuilder(account, {
  fee: "100",
  networkPassphrase: StellarSdk.Networks.TESTNET
})
  .addOperation(
    StellarSdk.Operation.beginSponsoringFutureReserves({
      sponsoredId: account.accountId(),
      source: masterKey.publicKey()
    })
  )
  .addOperation(
    StellarSdk.Operation.accountMerge({ destination: destKey.publicKey() }),
  ).addOperation(
    StellarSdk.Operation.createClaimableBalance({
      amount: "10",
      asset: StellarSdk.Asset.native(),
      claimants: [
        new StellarSdk.Claimant(account.accountId())
      ]
    }),
  ).addOperation(
    StellarSdk.Operation.claimClaimableBalance({
      balanceId: "00000000da0d57da7d4850e7fc10d2a9d0ebc731f7afb40574c03395b17d49149b91f5be",
    }),
  ).addOperation(
    StellarSdk.Operation.endSponsoringFutureReserves({
    })
  ).addOperation(
    StellarSdk.Operation.revokeAccountSponsorship({
      account: account.accountId(),
    })
  ).addOperation(
      StellarSdk.Operation.revokeTrustlineSponsorship({
        account: account.accountId(),
        asset: usd,
      })
  ).addOperation(
    StellarSdk.Operation.revokeOfferSponsorship({
      seller: account.accountId(),
      offerId: '12345'
    })
  ).addOperation(
    StellarSdk.Operation.revokeDataSponsorship({
      account: account.accountId(),
      name: 'foo'
    })
  ).addOperation(
    StellarSdk.Operation.revokeClaimableBalanceSponsorship({
      balanceId: "00000000da0d57da7d4850e7fc10d2a9d0ebc731f7afb40574c03395b17d49149b91f5be",
    })
  ).addOperation(
    StellarSdk.Operation.revokeSignerSponsorship({
      account: account.accountId(),
      signer: {
        ed25519PublicKey: sourceKey.publicKey()
      }
    })
  ).addOperation(
    StellarSdk.Operation.revokeSignerSponsorship({
      account: account.accountId(),
      signer: {
        sha256Hash: "da0d57da7d4850e7fc10d2a9d0ebc731f7afb40574c03395b17d49149b91f5be"
      }
    })
  ).addOperation(
    StellarSdk.Operation.revokeSignerSponsorship({
      account: account.accountId(),
      signer: {
        preAuthTx: "da0d57da7d4850e7fc10d2a9d0ebc731f7afb40574c03395b17d49149b91f5be"
      }
    })
  ).build();
```

### CODE REVIEW

**Review by vcarl — APPROVED** [review]
This seems good to me, but I'm not fully up to date on CAP-33. Nothing stands out as questionable

File: src/operations/end_sponsoring_future_reserves.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import xdr from '../generated/stellar-xdr_generated';
+
+/**
+ * Create a end sponsoring future reserves operation.
```
Comment (ire-and-curses) [inline src/operations/end_sponsoring_future_reserves.js:4]:
```suggestion
 * Create an "end sponsoring future reserves" operation.
```

File: src/operations/begin_sponsoring_future_reserves.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import xdr from '../generated/stellar-xdr_generated';
+import { StrKey } from '../strkey';
+import { Keypair } from '../keypair';
+
+/**
+ * Create a begin sponsoring future reserves operation.
```
Comment (ire-and-curses) [inline src/operations/begin_sponsoring_future_reserves.js:6]:
```suggestion
 * Create a "begin sponsoring future reserves" operation.
```

File: src/operations/revoke_sponsorship.js [hunk 3]
```diff
@@ -0,0 +1,274 @@
+import isString from 'lodash/isString';
+import xdr from '../generated/stellar-xdr_generated';
+import { StrKey } from '../strkey';
+import { Keypair } from '../keypair';
+import { Asset } from '../asset';
+
+/**
+ * Create a revoke sponsorship operation for an account.
```
Comment (ire-and-curses) [inline src/operations/revoke_sponsorship.js:8]:
```suggestion
 * Create a "revoke sponsorship" operation for an account.
```

File: src/operations/revoke_sponsorship.js [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+  const op = xdr.RevokeSponsorshipOp.revokeSponsorshipLedgerEntry(ledgerKey);
+  const opAttributes = {};
+  opAttributes.body = xdr.OperationBody.revokeSponsorship(op);
+  this.setSourceAccount(opAttributes, opts);
+
+  return new xdr.Operation(opAttributes);
+}
+
+/**
+ * Create a revoke sponsorship operation for a trustline.
```
Comment (ire-and-curses) [inline src/operations/revoke_sponsorship.js:42]:
```suggestion
 * Create a "revoke sponsorship" operation for a trustline.
```

File: src/operations/revoke_sponsorship.js [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+  const op = xdr.RevokeSponsorshipOp.revokeSponsorshipLedgerEntry(ledgerKey);
+  const opAttributes = {};
+  opAttributes.body = xdr.OperationBody.revokeSponsorship(op);
+  this.setSourceAccount(opAttributes, opts);
+
+  return new xdr.Operation(opAttributes);
+}
+
+/**
+ * Create a revoke sponsorship operation for an offer
```
Comment (ire-and-curses) [inline src/operations/revoke_sponsorship.js:85]:
```suggestion
 * Create a "revoke sponsorship" operation for an offer.
```

File: src/operations/revoke_sponsorship.js [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+  const op = xdr.RevokeSponsorshipOp.revokeSponsorshipLedgerEntry(ledgerKey);
+  const opAttributes = {};
+  opAttributes.body = xdr.OperationBody.revokeSponsorship(op);
+  this.setSourceAccount(opAttributes, opts);
+
+  return new xdr.Operation(opAttributes);
+}
+
+/**
+ * Create a revoke sponsorship operation for a data entry.
```
Comment (ire-and-curses) [inline src/operations/revoke_sponsorship.js:125]:
```suggestion
 * Create a "revoke sponsorship" operation for a data entry.
```

File: src/operations/revoke_sponsorship.js [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+  const op = xdr.RevokeSponsorshipOp.revokeSponsorshipLedgerEntry(ledgerKey);
+  const opAttributes = {};
+  opAttributes.body = xdr.OperationBody.revokeSponsorship(op);
+  this.setSourceAccount(opAttributes, opts);
+
+  return new xdr.Operation(opAttributes);
+}
+
+/**
+ * Create a revoke sponsorship operation for a claimable balance.
```
Comment (ire-and-curses) [inline src/operations/revoke_sponsorship.js:165]:
```suggestion
 * Create a "revoke sponsorship" operation for a claimable balance.
```

File: src/operations/revoke_sponsorship.js [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+  const op = xdr.RevokeSponsorshipOp.revokeSponsorshipLedgerEntry(ledgerKey);
+  const opAttributes = {};
+  opAttributes.body = xdr.OperationBody.revokeSponsorship(op);
+  this.setSourceAccount(opAttributes, opts);
+
+  return new xdr.Operation(opAttributes);
+}
+
+/**
+ * Create a revoke sponsorship operation for a signer.
```
Comment (ire-and-curses) [inline src/operations/revoke_sponsorship.js:199]:
```suggestion
 * Create a "revoke sponsorship" operation for a signer.
```

**Review by ire-and-curses — APPROVED** [review]
Only checked docstrings, looks good

**Pull request merged**